### PR TITLE
add basic electron test

### PR DESCRIPTION
**Pull request by pinussilvestrus**
To be rebased with #43 

-----

Currently not working 

* on Linux
* do not closing electron instance after tests on macOS

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,12 +14,15 @@
     "execa": "4.0.0",
     "husky": "4.2.3",
     "lint-staged": "10.0.8",
+    "mocha": "7.1.0",
     "mri": "1.1.4",
+    "npm-run-all": "4.1.5",
```
Comment (lipis) [inline package.json]: let's stay away from the npm-run-all, there is no real benefit.. and makes it harder when copying tasks directly to command line

Comment (pinussilvestrus) [reply package.json]: > no real benefit

I can see a benefit here

* the tasks tend to grow and grow and this composure with `yarn <task> & yarn <task> & yarn ...` will be clearly unreadable in some point
* npm-run-all gives us the chance to execute tasks in parallel in future (`run-p`)

Comment (lipis) [reply package.json]: > the tasks tend to grow and grow and this composure with `yarn <task> & yarn <task> & yarn ...` will be clearly unreadable in some point

Yet another reminder to keep them to the minimum

> npm-run-all gives us the chance to execute tasks in parallel in future (`run-p`)

If we ever going to gain anything out of we can rethink.. I would rather have only one way to run tasks.. using `yarn`

Comment (pinussilvestrus) [reply package.json]: alrighty

**Review by Sasha06865 — APPROVED** [review]

**Comment by pinussilvestrus** [comment]
I would love to see that one merged to start testing additional Electron functionality, e.g. introduced with #100 ... Still, some stuff to be fixed (especially on Linux, but I have no device to test that locally).

**Comment by shrirambalaji** [comment]
@pinussilvestrus I can help out with unit tests on Linux, as that's my daily driver. 

I'm not very clear on what needs to be tested here though. Would be great if you could point me on what to test.

**Comment by pinussilvestrus** [comment]
Thanks for your help @Shriram-Balaji !

Its basically this [failing test case](https://github.com/excalidraw/excalidraw-desktop/pull/46/checks?check_run_id=580609630#step:3:39).

This PRs only aims to add a basic test suite with a simple [starting test](https://github.com/excalidraw/excalidraw-desktop/pull/46/files#diff-6281935d66d4b753c6220d6db5669ab1R18), which currently fails on Linux.

**Comment by shrirambalaji** [comment]
@pinussilvestrus After some hiccups, I've got the unit tests working. So the tests were not working on CI/CD because when `Spectron` spins up a Chrome Instance to run the Selenium WebDriver, the internal ChromeDriver looks for an actual GUI. 

In server environments, we don't have a GUI and hence have to use `Xvfb` which emulates an actual X-Window Frame Buffer, using virtual memory. NOTE: This applies only to Linux, for Windows and macOS this is not applicable.

I've also done some renaming to the scripts, because we have both `lint` and `tests` now.

**Comment by shrirambalaji** [comment]
@pinussilvestrus  Also, do you think we should use jest instead of mocha, to keep it consistent with excalidraw web?

**Comment by pinussilvestrus** [comment]
Cool stuff 👏 

> Also, do you think we should use jest instead of mocha, to keep it consistent with excalidraw web?

makes absolute sense I guess

**Comment by pinussilvestrus** [comment]
> This applies only to Linux, for Windows and macOS this is not applicable.

.. but we should ensure tests work properly on mac and windows, too. Which they did before / without adding the xvfb support, am I right?

**Comment by shrirambalaji** [comment]
> > This applies only to Linux, for Windows and macOS this is not applicable.
> 
> .. but we should ensure tests work properly on mac and windows, too.

Yes, we should add test workflows for macOS and Windows too. For Windows - there's a "screen" available to run even in headless mode, and works on CI. For macOS, I can't seem to find the right documentation, but in theory, it should work without us needing to install / configure Xvfb.

>  Which they did before / without adding the xvfb support, am I right?

Well no, the tests were run against only the `ubuntu-latest` image on github actions, and there were no workflows for Windows / Mac ( ie. we were actually not running tests for them). There are build steps, but not the `lint` workflow if im not wrong.

**Comment by shrirambalaji** [comment]
@pinussilvestrus So I'm still running into the same errors for both macOS and Windows (Added separate workflows to run tests on em') . Could you try running the tests on your local machine, and check if the Mac tests work? Afaik they should work.

**Comment by pinussilvestrus** [comment]
I have currently the problem to install `spectron` correctly

image

Do you have an idea?

**Comment by shrirambalaji** [comment]
@pinussilvestrus Can you try `yarn install --force`? I experienced issues with spectron as well, when I bumped the version.

**Comment by pinussilvestrus** [comment]
It does not work for me unfortunately. Maybe I can figure out a fix for that.

**Comment by shrirambalaji** [comment]
> It does not work for me unfortunately. Maybe I can figure out a fix for that.

Hmm, any luck with it @pinussilvestrus? Can you check if [this](https://github.com/electron-userland/spectron/issues/144#issuecomment-459768796) works for you? I doubt that it would, but worth a try I guess. The only other option I can think of is to remove `node_modules` and retry, but I think you might have already checked that.

Looks like there's some issue of spectron being compatible with electron v9, https://github.com/electron-userland/spectron/issues/610. The error reported seems to be on Mac, not sure if it's related.

I think it might be worth it to downgrade electron to v8.x, considering that it might be more stable.

**Comment by shrirambalaji** [comment]
@pinussilvestrus I have downgraded electron to 8.3.0, and along with that have gotten the tests running on both `macos` and `linux`. Also, I'm guessing if you clear `node_modules`, spectron should install on your local system now.

But, I can't seem to clearly figure out why, windows is not working. Would it be ok for us to add the `windows` test workflow later?

**Comment by pinussilvestrus** [comment]
I think that's ok for now. Can you maybe add an issue for adding windows unit test support as a next step? Then we would be ok with merging this one

**Comment by shrirambalaji** [comment]
> I think that's ok for now. Can you maybe add an issue for adding windows unit test support as a next step? Then we would be ok with merging this one

Just created an issue for that https://github.com/excalidraw/excalidraw-desktop/issues/163

**Pull request merged**